service/storage_service.cc

```diff
@@ -5503,6 +5503,10 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(raft
     try {
         auto& raft_server = _group0->group0_server();
         auto group0_holder = _group0->hold_group0_gate();
+
+        // make sure any pending group0 upgrade is completed
+        co_await _group0->client().wait_until_group0_upgraded(_group0_as);
+
```

Comment:
Can you explain what is the race condition exactly? Also I do not understand why is this suppose to fix it. `aft_topology_cmd_handler` has nothing to do with group0 command application.

Reply:
The particular offender here is the call to `mark_existing_views_as_built()`: https://github.com/scylladb/scylladb/blob/3a6c0a9b36b38a8f2a8b777abc664c8eca8ec4b5/service/storage_service.cc#L5643-L5647

That call causes the raft `start_operation()` to be called (see the callstack here: https://github.com/scylladb/scylladb/issues/20844#issuecomment-2470269750) and the issue happens when the group0 upgrade is in progress at that time (so the `start_operation()` sees the status `group0_upgrade_state::synchronize` and that triggers the exception).

Adding the `wait_until_group0_upgraded()` makes sure that the eventual group0 upgrade in progress is finished before we continue. The question is also what is the right place to put this wait - initially had it just before the `mark_existing_views_as_built()` to fix that particular case, but after discussion with Kamil moved it at the beginning of the whole raft handler to apply to any raft command.

The `raft_topology_cmd::command::stream_ranges` command is being triggered on the other node (in the topology coordinator), which seems to not take the upgrade status of the other nodes into account (maybe the question is whether it should, but perhaps it's simpler this way).

Reply:
But it looks like we try to fix symptoms here, not the root cause. The node should not suppose to get any rpc cmd commands or raft log commands until it knows the mode it operates in (or at least that what I would expect during upgrade). Here https://github.com/scylladb/scylladb/issues/21114#issuecomment-2440080855 the problem is very similar to what you describe  except that the mode is not known during raft command application not topology rpc. 

Also the commit message is incorrect. It says "group0 commands to be applied during the group0 upgrade procedure", but here you protect against topology cmd command being executed while the mode is not known.